File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Permutation Sequence/Java/Solution.java
import java.util.*;

class Solution {
    public String getPermutation(int n, int k) {
        List<Integer> nums = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            nums.add(i);
        }
        
        return backTracking(n, k - 1, nums, new StringBuilder());
    }

    private String backTracking(int n, int k, List<Integer> nums, StringBuilder sb) {
        if (nums.size() == 1) {
            return sb.append(nums.get(0)).toString();
        }

        int seqCount = calculateFactorial(n - 1);
        int target = k / seqCount;

        sb.append(nums.get(target));
        nums.remove(target);
        return backTracking(n - 1, k % seqCount, nums, sb);
    }

    private int calculateFactorial(int n) {
        int result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }

        return result;
    }
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Permutation Sequence/solution.py
class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        nums = [i for i in range(1, n + 1)]
        return self.recursive(nums, k - 1, "")
    
    def recursive(self, nums, k, answer):
        if (len(nums) == 1):
            return answer + str(nums[0])
        
        case_count = self.calculate_factorial(len(nums) - 1)
        target = k // case_count
        answer += str(nums[target])
        nums.pop(target)

        return self.recursive(nums, k % case_count, answer)
    
    def calculate_factorial(self, n):
        result = 1
        for i in range(1, n + 1):
            result *= i
        
        return result
